tephra-core/src/main/java/co/cask/tephra/distributed/TransactionServiceThriftClient.java

```diff
@@ -57,9 +57,14 @@ public ByteBuffer apply(byte[] input) {
   TTransactionServer.Client client;
 
   /**
+   * Whether this client is valid for use.
+   */
+  private boolean isValid = true;
```

Comment:
It might be a good idea to make this `AtomicBoolean`, as the `isValid()` method call can happen in a different thread than the one that executes the transaction (although it might not be the case now).